### PR TITLE
Log a warning when cluster admins enable Insecure TLS Ciphers for kubelet or apiserver

### LINKED ISSUE #81145: TOB-K8S-037: Kubelet supports insecure TLS ciphersuites
This issue was reported in the [Kubernetes Security Audit Report](https://github.com/kubernetes/community/blob/master/wg-security-audit/findings/Kubernetes%20Final%20Report.pdf)

**Description**
Kubelet allows administrators to configure TLS connections to use a variety of insecure cipher suites. In particular it supports the use of RC4 and 3DES in its symmetric suites. RC4 has known bias in its output and should never be used, while 3DES is an extremely deprecated 64-bit block cipher which is both slow and unneeded. Additionally, non-forward secure key exchange is supported (TLS_RSA_*). This, along with SHA-based cipher suites, should be deprecated and replaced.

**Recommendation**
Remove support for any cipher suite that uses RC4 or 3DES as well as non-forward secure key exchange suites (TLS_RSA_*). Deprecate all but the following cipher suite options for TLS versions up through 1.2:

TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256

Use the following cipher suites for TLS 1.3:

TLS_AES_128_GCM_SHA256 
TLS_CHACHA20_POLY1305_SHA256
TLS_AES_256_GCM_SHA384

**Anything else we need to know?**:

See #81146 for current status of all issues created from these findings.

The vendor gave this issue an ID of TOB-K8S-037 and it was finding 37 of the report.

The vendor considers this issue Informational Severity.

To view the original finding, begin on page 89 of the [Kubernetes Security Review Report](https://github.com/kubernetes/community/blob/master/wg-security-audit/findings/Kubernetes%20Final%20Report.pdf)

**Environment**:

- Kubernetes version: 1.13.4

### PR DESCRIPTION
**Pull request by pjbgf**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Kubernetes allows cluster admins to select what TLS ciphers to support on Kubelet and/or API Server. This PR logs a warning message when any chosen cipher is deemed insecure based of [tls.InsecureCipherSuites()](https://golang.org/src/crypto/tls/cipher_suites.go#L81).

**Which issue(s) this PR fixes**:

Part of #81145

**Special notes for your reviewer**:
The list of ciphers below will now be marked as insecure:
```
TLS_RSA_WITH_RC4_128_SHA
TLS_RSA_WITH_AES_128_CBC_SHA256
TLS_ECDHE_ECDSA_WITH_RC4_128_SHA
TLS_ECDHE_RSA_WITH_RC4_128_SHA
TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256
TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256
```
Once the Kubernetes moves to golang 1.14, this list will be loaded from [tls.InsecureCipherSuites()](https://golang.org/src/crypto/tls/cipher_suites.go#L81).

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

/sig api-machinery
/sig auth
/area security
/priority backlog

### CODE REVIEW

**Comment by neolit123** [comment]
> Fixes #81145

i personally slightly disagree with the argument in the above issue, because the same argument applies to everything out there that provides insecure cyphers including sshd or golang which k8s uses as the algorithm provider in this case.

adding a feature gate feels like an extra burden for us to carry.
also note that feature gates have a lifecycle, must graduate to GA at some point (removing the legacy cypher support), be tracked per release by the k8s release team. ideally this should have been logged as a kubernetes/enhancements ticket + a KEP for it should have been written that provides documentation and testing details.

**Comment by pjbgf** [comment]
@neolit123 thank you for taking the time to comment.

> > Fixes #81145
> 
> i personally slightly disagree with the argument in the above issue, because the same argument applies to everything out there that provides insecure cyphers including sshd or golang which k8s uses as the algorithm provider in this case.

I am not sure I understand what you mean. My thought the scope of #81145 was around the TLS communication amongst the kubernetes components - which this PR fixes. But I would not be surprised if I am wrong on that one. :) 

Should I remove the "Fixes" so it does not close the original issue?

> adding a feature gate feels like an extra burden for us to carry.
> also note that feature gates have a lifecycle, must graduate to GA at some point (removing the legacy cypher support), be tracked per release by the k8s release team. ideally this should have been logged as a kubernetes/enhancements ticket + a KEP for it should have been written that provides documentation and testing details.

That makes a lot of sense - I will remove the feature gate. After all, the recommended ciphers to be supported are 9 in total. 6 of them have been supported by Kubernetes for 3 years now, which is good enough to comply with the Kubernetes version skew support policy. 

The other 3 are TLS 1.3 ciphers, which will only be supported from v1.19 anyway, as they were [merged into master](https://github.com/kubernetes/kubernetes/commit/34226de021dce74cb4e1677c4f4d0f2350fa2253) today. 

Just for reference, here's the commit that added the 6 ciphers mentioned above: https://github.com/kubernetes/kubernetes/commit/d7dbc96c70d480f0b81cd83ae3abd34b69c1e70d

**Comment by neolit123** [comment]
> I am not sure I understand what you mean. 

the issue in question talks about removing support for the insecure ciphers, which i said i slightly disagree with, because ultimately users are already allowed to control what suites are allowed via `--tls-cipher-suites`.

> That makes a lot of sense - I will remove the feature gate. After all, the recommended ciphers to be supported are 9 in total. 6 of them have been supported by Kubernetes for 3 years now, which is good enough to comply with the Kubernetes version skew support policy.

i'd wait for a response from SIG Auth (@kubernetes/sig-auth-pr-reviews )
before doing that.

**Comment by liggitt** [comment]
I don't think a feature gate around an opt-in config makes sense.

go1.14 adds stdlib support for identifying cipher suites considered insecure:

> The new CipherSuites and InsecureCipherSuites functions return a list of currently implemented cipher suites. The new CipherSuiteName function returns a name for a cipher suite ID.

Once we update to go1.14, I'd like to see warnings emitted if someone uses `--tls-cipher-suites` to enable a cipher suite considered insecure, but that is all I would recommend doing.

**Comment by pjbgf** [comment]
> go1.14 adds stdlib support for identifying cipher suites considered insecure:
> 
> > The new CipherSuites and InsecureCipherSuites functions return a list of currently implemented cipher suites. The new CipherSuiteName function returns a name for a cipher suite ID.

@liggitt that makes all the sense, thank you for providing direction. 
I have changed the scope of this PR to simply remove the insecure ciphers - as per [tls.InsecureCipherSuites()](https://golang.org/src/crypto/tls/cipher_suites.go#L81). This way Kubernetes users will get the benefit right away. 

On a side note, #81145 implicitly asks to remove both insecure and weak ciphers. However, tls.InsecureCipherSuites() only contains insecure ciphers. Is it worth:

1) Mastering the information of weak ciphers within Kubernetes (following the info from the audit), and use it in combination with the tls.InsecureCiphersSuites() to remove/warn users about both cases.
2) Update the issue explaining we will only focus on insecure ciphers for the time being, and perhaps close the issue.

Or perhaps another option I did not think of?

> Once we update to go1.14, I'd like to see warnings emitted if someone uses --tls-cipher-suites to enable a cipher suite considered insecure, but that is all I would recommend doing

Was thinking on doing this as a separate PR. Does that make sense or should I just add to this one?

**Comment by liggitt** [comment]
I would limit this change to:
* warning if an item included in InsecureCipherSuites is explicitly opted into by a cluster admin
* partioning the list of accepted cipher suites in the flag help to clarify which ones are considered insecure by the stdlib

**Comment by pjbgf** [comment]
> I would limit this change to:
> 
> * warning if an item included in InsecureCipherSuites is explicitly opted into by a cluster admin
> * partioning the list of accepted cipher suites in the flag help to clarify which ones are considered insecure by the stdlib

@liggitt I made the changes accordingly. PTAL

File: staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go [hunk 1]
```diff
@@ -53,21 +48,65 @@ var ciphers = map[string]uint16{
 	"TLS_AES_256_GCM_SHA384":                  tls.TLS_AES_256_GCM_SHA384,
 }
 
-func TLSCipherPossibleValues() []string {
```
Comment (liggitt) [inline staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go:56]: preserve this and return the combination of both lists

File: staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag_test.go [hunk 2]
```diff
@@ -86,18 +86,18 @@ func TestConstantMaps(t *testing.T) {
 		if strings.HasPrefix(declName, "VersionTLS") {
 			discoveredVersions[declName] = true
 		}
-		if strings.HasPrefix(declName, "TLS_RSA_") || strings.HasPrefix(declName, "TLS_ECDHE_") ||
-			strings.HasPrefix(declName, "TLS_AES_") || strings.HasPrefix(declName, "TLS_CHACHA20_") {
```
Comment (liggitt) [inline staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag_test.go:90]: this drops ciphers that were just added in https://github.com/kubernetes/kubernetes/pull/90843, please revert this change

File: cmd/kubelet/app/server.go [hunk 3]
```diff
@@ -990,6 +991,17 @@ func InitializeTLS(kf *options.KubeletFlags, kc *kubeletconfiginternal.KubeletCo
 		return nil, err
 	}
 
+	if len(tlsCipherSuites) > 0 {
+		insecureCiphers := flag.InsecureTLSCiphers()
+		for i := 0; i < len(tlsCipherSuites); i++ {
+			for cipherName, cipherID := range insecureCiphers {
+				if tlsCipherSuites[i] == cipherID {
+					klog.Warningf("[kubelet] WARNING: Use of insecure cipher '%s' detected.", cipherName)
```
Comment (liggitt) [inline cmd/kubelet/app/server.go:999]:
```suggestion
					klog.Warningf("Use of insecure cipher '%s' detected.", cipherName)
```

File: staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go (hunk 1 above)
Comment (pjbgf) [reply staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go:56]: done, ptal

File: staging/src/k8s.io/apiserver/pkg/server/secure_serving.go [hunk 4]
```diff
@@ -56,6 +57,14 @@ func (s *SecureServingInfo) tlsConfig(stopCh <-chan struct{}) (*tls.Config, erro
 	}
 	if len(s.CipherSuites) > 0 {
 		tlsConfig.CipherSuites = s.CipherSuites
+		insecureCiphers := flag.InsecureTLSCiphers()
+		for i := 0; i < len(s.CipherSuites); i++ {
+			for cipherName, cipherID := range insecureCiphers {
+				if s.CipherSuites[i] == cipherID {
+					klog.Warningf("[api-server] WARNING: Use of insecure cipher '%s' detected.", cipherName)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/secure_serving.go:64]:
```suggestion
					klog.Warningf("Use of insecure cipher '%s' detected.", cipherName)
```

File: staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go [hunk 5]
```diff
@@ -53,21 +48,76 @@ var ciphers = map[string]uint16{
...
+// PreferredTLSCipherNames returns a list of cipher suite names implemented by crypto/tls.
+func PreferredTLSCipherNames() []string {
 	cipherKeys := sets.NewString()
 	for key := range ciphers {
 		cipherKeys.Insert(key)
 	}
 	return cipherKeys.List()
 }
 
+func tlsCipherPossibleValues() map[string]uint16 {
```
Comment (liggitt) [inline staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go:90]: nit: rename to allCiphers

File: cluster/images/conformance/go-runner/BUILD [hunk 6]
```diff
@@ -41,6 +41,7 @@ go_test(
         "env_test.go",
         "tar_test.go",
     ],
+    data = glob(["testdata/**"]),
```
Comment (liggitt) [inline cluster/images/conformance/go-runner/BUILD:44]: revert this?

Comment (pjbgf) [reply cluster/images/conformance/go-runner/BUILD]: hack/update-bazel.sh created it - I wasn't quite sure why. Just removed it.

**Comment by pjbgf** [comment]
Re-based to fix conflict.
